report/renderer/swap-locale-feature.js

```diff
@@ -0,0 +1,109 @@
+ */
+'use strict';
+
+/* eslint-env browser */
+
+/** @typedef {import('./dom.js').DOM} DOM */
+/** @typedef {import('./report-ui-features').ReportUIFeatures} ReportUIFeatures */
+/** @typedef {import('../../lighthouse-core/lib/i18n/locales').LhlMessages} LhlMessages */
+
+export class SwapLocaleFeature {
+  /**
+   * @param {ReportUIFeatures} reportUIFeatures
+   * @param {DOM} dom
+   * @param {{i18nModuleSrc: string, fetchData: (localeModuleName: string) => Promise<LhlMessages|undefined>}} options
+   *        Specifiy the URL where the i18n module script can be found, and the URLS for the locale JSON files.
+   */
+  constructor(reportUIFeatures, dom, options) {
+    this._reportUIFeatures = reportUIFeatures;
+    this._dom = dom;
+    this._swapLocaleOptions = options;
+  }
+
+  async enable() {
+    try {
+      await this._enable();
+    } catch (err) {
+      // eslint-disable-next-line no-console
+      console.error('failed to enable swap locale feature', err);
+    }
+  }
+
+  async _enable() {
+    if (!this._reportUIFeatures.json.i18n.icuMessagePaths) {
+      throw new Error('missing icuMessagePaths');
+    }
+
+    const i18nModule = await this._getI18nModule();
+    const currentLocale = this._reportUIFeatures.json.configSettings.locale;
+
+    const containerEl = this._dom.find('.lh-tools-locale__selector-wrapper', this._dom.document());
+    const inputEl = this._dom.createChildOf(containerEl, 'select', 'lh-locale-selector');
+    inputEl.setAttribute('type', 'text');
+    inputEl.name = 'lh-locale-list';
+
+    const toggleEl = this._dom.find('.lh-tool-locale__button', this._dom.document());
+    toggleEl.addEventListener('click', () => {
+      toggleEl.classList.toggle('lh-active');
+    });
+
+    for (const locale of i18nModule.availableLocales) {
+      const optionEl = this._dom.createChildOf(inputEl, 'option', '');
+      optionEl.value = locale;
+      optionEl.textContent = locale;
+      if (locale === currentLocale) optionEl.selected = true;
+
+      // @ts-ignore
+      if (window.Intl && window.Intl.DisplayNames) {
+        // @ts-ignore
+        const currentLocaleDisplay = new Intl.DisplayNames([currentLocale], {type: 'language'});
+        // @ts-ignore
+        const optionLocaleDisplay = new Intl.DisplayNames([locale], {type: 'language'});
+
+        const optionLocaleName = optionLocaleDisplay.of(locale);
+        const currentLocaleName = currentLocaleDisplay.of(locale);
+        if (optionLocaleName !== currentLocaleName) {
+          optionEl.textContent = `${optionLocaleName} – ${currentLocaleName}`;
+        } else {
+          optionEl.textContent = currentLocaleName;
+        }
+      }
+    }
+
+    inputEl.addEventListener('change', () => {
+      const locale = /** @type {LH.Locale} */ (inputEl.value);
+      this._swapLocale(locale);
+    });
+  }
+
+  /**
+   * @param {LH.Locale} locale
+   */
+  async _swapLocale(locale) {
```

Comment:
I'm thinking that PSI/LB/whatever could call this directly (avoiding the enable which adds stuff to the UI). so:

```js
const feat = new SwapLocalesFeature(reportUIFeatures, dom, {
  async fetchData() { ... }
});
await fetch._swapLocale('blah');

// listening to refresh event somewhere else ...
```